pkg/controller/function/function_controller.go

```diff
@@ -83,62 +107,178 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
+// +kubebuilder:rbac:groups="",resources=serviceaccounts,verbs=get;watch;update;list
+// +kubebuilder:rbac:groups=";apps;extensions",resources=deployments,verbs=create;get;watch;update;delete;list;update;patch
 func (r *ReconcileFunction) Reconcile(request reconcile.Request) (reconcile.Result, error) {
 
-	fnConfigName := "fn-config"
-	fnConfigNamespace := "default"
-	fnConfigNameEnv := os.Getenv("CONTROLLER_CONFIGMAP")
-	fnConfigNamespaceEnv := os.Getenv("CONTROLLER_CONFIGMAP_NS")
+	// Get Function Controller Configuration
+	fnConfig := &corev1.ConfigMap{}
+	if err := r.getFunctionControllerConfiguration(fnConfig); err != nil {
+		return reconcile.Result{}, err
+	}
 
-	if len(fnConfigNameEnv) > 0 {
-		fnConfigName = fnConfigNameEnv
+	// Get a new *RuntimeInfo
+	rnInfo, err := runtimeUtil.New(fnConfig)
+	if err != nil {
+		log.Error(err, "Error while trying to get a new RuntimeInfo instance", "namespace", fnConfig.Namespace, "name", fnConfig.Name)
+		return reconcile.Result{}, err
 	}
 
-	if len(fnConfigNamespaceEnv) > 0 {
-		fnConfigNamespace = fnConfigNamespaceEnv
+	// Get Function instance
+	fn := &runtimev1alpha1.Function{}
+	if err := r.getFunctionInstance(request, fn); err != nil {
+		// status of the functon must change to error.
+		r.updateFunctionStatus(fn, runtimev1alpha1.FunctionConditionError)
+
+		log.Error(err, "Error reading Function instance", "namespace", request.Namespace, "name", request.Name)
+		return reconcile.Result{}, err
 	}
-	fnConfig := &corev1.ConfigMap{}
-	err := r.Get(context.TODO(), types.NamespacedName{Name: fnConfigName, Namespace: fnConfigNamespace}, fnConfig)
 
-	if err != nil {
-		log.Error(err, "Unable to read Function controller config: %v from Namespace: %v", fnConfigName, fnConfigNamespace)
+	// Create Function's ConfigMap
+	foundCm := &corev1.ConfigMap{}
+	deployCm := &corev1.ConfigMap{}
+	if err := r.createFunctionConfigMap(foundCm, deployCm, fn); err != nil {
+		// status of the functon must change to error.
+		r.updateFunctionStatus(fn, runtimev1alpha1.FunctionConditionError)
+
+		log.Error(err, "Error while trying to create the Function's ConfigMap", "namespace", deployCm.Namespace, "name", deployCm.Name)
 		return reconcile.Result{}, err
 	}
 
-	rnInfo, err := runtimeUtil.New(fnConfig)
-	if err != nil {
-		fmt.Printf("Error in reading ConfigMap: %v", err)
+	// Update Function's ConfigMap
+	if err := r.updateFunctionConfigMap(foundCm, deployCm); err != nil {
+		// status of the functon must change to error.
+		r.updateFunctionStatus(fn, runtimev1alpha1.FunctionConditionError)
+
+		log.Error(err, "Error while trying to update Function's ConfigMap:", "namespace", deployCm.Namespace, "name", deployCm.Name)
 		return reconcile.Result{}, err
 	}
-	// Fetch the Function instance
-	fn := &runtimev1alpha1.Function{}
-	err = r.Get(context.TODO(), request.NamespacedName, fn)
+
+	// Create function's image name
+	hash := sha256.New()
+	print("foundCM: %v", foundCm)
```

Comment:
debug message ... please log as debug

Reply:
I will removed. That was a flag for debugging.